File: src/java/magmac/app/lang/web/TypescriptLang.java
package magmac.app.lang.web;

import magmac.api.Option;
import magmac.api.collect.list.List;
import magmac.app.compile.error.CompileResult;
import magmac.app.compile.node.MapNode;
import magmac.app.compile.node.Node;
import magmac.app.lang.CommonLang;
import magmac.app.lang.Destructors;
import magmac.app.lang.Serializable;
import magmac.app.lang.common.Annotation;
import magmac.app.lang.java.JavaLang;
import magmac.app.lang.node.AbstractReturnNode;
import magmac.app.lang.node.Conditional;
import magmac.app.lang.node.ConditionalType;
import magmac.app.lang.node.Modifier;
import magmac.app.lang.node.Operator;
import magmac.app.lang.node.ParameterizedMethodHeader;
import magmac.app.lang.node.PostVariant;
import magmac.app.lang.node.Segment;
import magmac.app.lang.node.StructureValue;

public final class TypescriptLang {
    public interface Argument extends Serializable {
    }

    public interface TypeScriptParameter extends Serializable {
    }

    public interface TypescriptBlockHeader extends Serializable {
    }

    public interface TypescriptStructureMember extends Serializable {
    }

    public interface TypeScriptMethodHeader extends Serializable {
    }

    public interface Type extends Serializable {
    }

    public interface TypeScriptRootSegment extends Serializable {
    }

    public interface FunctionSegment extends Serializable {
        interface Value extends Serializable {
        }
    }

    public interface Value extends Caller, Argument, Assignable {
    }

    public interface Assignable extends Serializable {
    }

    public record Number(String value) implements Value {
        @Override
        public Node serialize() {
            return new MapNode("number").withString("value", this.value);
        }
    }

    public static final class StructureNode implements TypeScriptRootSegment {
        private final StructureType type;
        private final StructureValue<Type, TypescriptStructureMember> value;

        public StructureNode(StructureType type, StructureValue<Type, TypescriptStructureMember> structureNode) {
            this.type = type;
            this.value = structureNode;
        }

        private Node serializeImplementsParams() {
            return this.value.maybeImplemented()
                    .map(implemented -> new MapNode().withNodeListSerialized("implemented", implemented))
                    .orElse(new MapNode());
        }

        private Node serializeExtendedParams() {
            return this.value.maybeExtended()
                    .map(extended -> new MapNode().withNodeListSerialized("extended", extended))
                    .orElse(new MapNode());
        }

        private Node serializeTypeParams() {
            return this.value.maybeTypeParams()
                    .map(typeParams -> new MapNode().withNodeListSerialized("type-parameters", typeParams))
                    .orElse(new MapNode());
        }

        @Override
        public Node serialize() {
            return new MapNode(this.type.text())
                    .withString("name", this.value.name())
                    .withNodeListAndSerializer("modifiers", this.value.modifiers(), Serializable::serialize)
                    .withNodeListAndSerializer("members", this.value.members(), Serializable::serialize)
                    .merge(this.serializeTypeParams())
                    .merge(this.serializeExtendedParams())
                    .merge(this.serializeImplementsParams());
        }
    }

    public record TypeScriptImport(List<Segment> values, List<Segment> segments) implements TypeScriptRootSegment {
        @Override
        public Node serialize() {
            return new MapNode("import")
                    .withNodeListAndSerializer("values", this.values, Segment::serialize)
                    .withNodeListAndSerializer("segments", this.segments, Segment::serialize);
        }
    }

    public record TypescriptRoot(List<TypeScriptRootSegment> children) implements Serializable {
        @Override
        public Node serialize() {
            return new MapNode("root")
                    .withNodeListAndSerializer("children", this.children, TypeScriptRootSegment::serialize);
        }
    }

    public record TypescriptMethod(
            ParameterizedMethodHeader<TypeScriptParameter> header,
            Option<List<FunctionSegment>> maybeChildren
    ) implements TypescriptStructureMember {
        @Override
        public Node serialize() {
            var node = new MapNode("method")
                    .withNodeSerialized("header", this.header);

            return this.maybeChildren.map(children -> node.withNodeListSerialized("children", children)).orElse(node);
        }
    }

    public static final class TemplateType implements Type {
        private final JavaLang.Symbol base;
        private final Option<List<Type>> maybeTypeArguments;

        public TemplateType(JavaLang.Symbol base, Option<List<Type>> maybeTypeArguments) {
            this.base = base;
            this.maybeTypeArguments = maybeTypeArguments;
        }

        @Override
        public Node serialize() {
            var node = new MapNode("template")
                    .withNodeSerialized("base", this.base);

            return this.maybeTypeArguments.map(typeArguments -> {
                return node.withNodeListSerialized("arguments", typeArguments);
            }).orElse(node);
        }
    }

    public static final class TypescriptConditional extends Conditional<Value> implements TypescriptBlockHeader {
        public TypescriptConditional(ConditionalType type, Value condition) {
            super(type, condition);
        }

        @Override
        public Node serialize() {
            return new MapNode(this.type.text())
                    .withNodeSerialized("condition", this.condition);
        }
    }

    public static final class Block extends magmac.app.lang.node.Block<TypescriptBlockHeader, FunctionSegment> implements FunctionSegment {
        public Block(TypescriptBlockHeader header, List<FunctionSegment> segments) {
            super(header, segments);
        }

        @Override
        public Node serialize() {
            return new MapNode("block")
                    .withNodeSerialized("header", this.header)
                    .withNodeListSerialized("children", this.children);
        }
    }

    public static final class TypescriptConstructor implements TypeScriptMethodHeader {
        @Override
        public Node serialize() {
            return new MapNode("constructor");
        }
    }

    public static final class Definition extends CommonLang.AbstractDefinition<Type> implements TypeScriptParameter, Assignable, TypeScriptMethodHeader {
        public Definition(Option<List<Annotation>> maybeAnnotations, List<Modifier> modifiers, String name, Option<List<TypeParam>> maybeTypeParams, Type type) {
            super(maybeAnnotations, modifiers, name, maybeTypeParams, type);
        }

        @Override
        public Node serialize() {
            var node = new MapNode("definition")
                    .withString("name", this.name)
                    .withNodeSerialized("type", this.type);

            if (this.modifiers.isEmpty()) {
                return node;
            }
            else {
                return node.withNodeListSerialized("modifiers", this.modifiers);
            }
        }

        public Definition withModifier(Modifier modifier) {
            return new Definition(this.maybeAnnotations, this.modifiers.addLast(modifier), this.name, this.maybeTypeParams, this.type);
        }
    }

    public static final class ArrayType implements Type {
        private final Type childType;

        public ArrayType(Type arrayType) {
            this.childType = arrayType;
        }

        @Override
        public Node serialize() {
            return new MapNode("array").withNodeSerialized("child", this.childType);
        }
    }

    public record TypeParam(String value) implements Serializable {
        public static CompileResult<TypeParam> deserialize(Node node) {
            return Destructors.destruct(node)
                    .withString("value")
                    .complete(TypeParam::new);
        }

        @Override
        public Node serialize() {
            return new MapNode().withString("value", this.value);
        }
    }

    public record Construction(Type type) implements Caller {
        @Override
        public Node serialize() {
            return new MapNode("construction").withNodeSerialized("type", this.type);
        }
    }

    public record Post(PostVariant variant, Value value) implements FunctionSegment.Value {
        @Override
        public Node serialize() {
            return new MapNode(this.variant.type()).withNodeSerialized("child", this.value);
        }
    }

    public record Access(Value receiver, String property) implements Value {
        @Override
        public Node serialize() {
            return new MapNode("data-access")
                    .withNodeSerialized("receiver", this.receiver)
                    .withString("property", this.property);
        }
    }

    public static final class Break implements FunctionSegment.Value {
        @Override
        public Node serialize() {
            return new MapNode("break");
        }
    }

    public static final class Continue implements FunctionSegment.Value {
        @Override
        public Node serialize() {
            return new MapNode("continue");
        }
    }

    public static final class Return extends AbstractReturnNode<Value> implements FunctionSegment.Value, FunctionSegment {
        public Return(TypescriptLang.Value child) {
            super(child);
        }

        @Override
        public Node serialize() {
            return new MapNode("return").withNodeSerialized("child", this.child);
        }
    }

    /**
     * Invocation node for the TypeScript AST. An example syntax would be
     * {@code myFunc(1, "a")}. The {@code caller} field stores {@code myFunc}
     * and the {@code arguments} list holds {@code 1} and {@code "a"}.
     */
    public static class Invokable extends magmac.app.lang.java.Invokable<Caller, Argument> implements Value, FunctionSegment.Value {
        public Invokable(Caller caller, List<Argument> arguments) {
            super(caller, arguments);
        }

        @Override
        public Node serialize() {
            return new MapNode("invokable")
                    .withNodeSerialized("caller", this.caller)
                    .withNodeListSerialized("arguments", this.arguments);
        }
    }

    public record Char(String value) implements Value {
        @Override
        public Node serialize() {
            return new MapNode("char").withString("value", this.value);
        }
    }

    public record Index(Value parent, Value argument) implements Value {
        @Override
        public Node serialize() {
            return new MapNode("index").withNodeSerialized("parent", this.parent).withNodeSerialized("argument", this.argument);
        }
    }

    public record Not(Value child) implements Value {
        @Override
        public Node serialize() {
            return new MapNode("not").withNodeSerialized("child", this.child);
        }
    }

    public record StringValue(java.lang.String value) implements Value {
        @Override
        public Node serialize() {
            return new MapNode("string").withString("value", this.value);
        }
    }

    public record Symbol(String value) implements Value {
        @Override
        public Node serialize() {
            return new MapNode("symbol").withString("value", this.value);
        }
    }

    public record Operation(
            Value left,
            Operator operator,
            Value right
    ) implements Value {
        @Override
        public Node serialize() {
            return new MapNode(this.operator.type())
                    .withNodeSerialized("left", this.left)
                    .withNodeSerialized("right", this.right);
        }
    }

    public static class Whitespace implements
            TypeScriptRootSegment,
            TypescriptStructureMember,
            TypeScriptParameter,
            FunctionSegment,
            Argument {
        @Override
        public Node serialize() {
            return new MapNode("whitespace");
        }
    }

    public record Assignment(Assignable assignable, Value value) implements FunctionSegment.Value {
        @Override
        public Node serialize() {
            return new MapNode("assignment")
                    .withNodeSerialized("destination", this.assignable)
                    .withNodeSerialized("source", this.value);
        }
    }

    public enum StructureType {
        Class("class"),
        Enum("enum"),
        Interface("interface");

        private final String text;

        StructureType(String text) {
            this.text = text;
        }

        public String text() {
            return this.text;
        }
    }
}

File: src/web/magmac/app/lang/web/TypescriptLang.ts
import { Option } from "../../../../magmac/api/Option";
import { List } from "../../../../magmac/api/collect/list/List";
import { CompileResult } from "../../../../magmac/app/compile/error/CompileResult";
import { MapNode } from "../../../../magmac/app/compile/node/MapNode";
import { Node } from "../../../../magmac/app/compile/node/Node";
import { CommonLang } from "../../../../magmac/app/lang/CommonLang";
import { Destructors } from "../../../../magmac/app/lang/Destructors";
import { Serializable } from "../../../../magmac/app/lang/Serializable";
import { JavaLang } from "../../../../magmac/app/lang/java/JavaLang";
import { Block } from "../../../../magmac/app/lang/node/Block";
import { Conditional } from "../../../../magmac/app/lang/node/Conditional";
import { ConditionalType } from "../../../../magmac/app/lang/node/ConditionalType";
import { ParameterizedMethodHeader } from "../../../../magmac/app/lang/node/ParameterizedMethodHeader";
import { PostVariant } from "../../../../magmac/app/lang/node/PostVariant";
import { ReturnNode } from "../../../../magmac/app/lang/node/ReturnNode";
import { Segment } from "../../../../magmac/app/lang/node/Segment";
import { StructureValue } from "../../../../magmac/app/lang/node/StructureValue";
import { TypeArguments } from "../../../../magmac/app/lang/node/TypeArguments";
export class TypescriptWhitespace {
	serialize() : Node {return new MapNode( 0);;}
}
export interface TypescriptArgument {
}
export interface TypeScriptParameter {
}
export interface TypescriptBlockHeader {
}
export interface TypescriptStructureMember {
}
export interface TypescriptLambdaContent {
}
export interface TypeScriptMethodHeader {
}
export interface TypeScriptType {
}
export interface TypeScriptRootSegment {
}
export interface TypescriptFunctionSegment {
}
export interface TypescriptValue {
}
export class Number {
	serialize() : Node {return new MapNode( 0).withString( 0, 0.value);;}
}
export class TypescriptStructureNode {
	TypescriptStructureNode(type : TypescriptStructureType, structureNode : StructureValue<TypeScriptType, TypescriptStructureMember>) : public {break;break;;}
	serializeImplementsParams() : Node {return 0.value.maybeImplemented( ).map( 0).orElse( new MapNode( ));;}
	serializeExtendedParams() : Node {return 0.value.maybeExtended( ).map( 0).orElse( new MapNode( ));;}
	serializeTypeParams() : Node {return 0.value.maybeTypeParams( ).map( 0).orElse( new MapNode( ));;}
	serialize() : Node {return new MapNode( 0.type.name( ).toLowerCase( )).withString( 0, 0.value.name( )).withNodeListAndSerializer( 0, 0.value.modifiers( ), 0.serialize).withNodeListAndSerializer( 0, 0.value.members( ), 0.serialize).merge( 0.serializeTypeParams( )).merge( 0.serializeExtendedParams( )).merge( 0.serializeImplementsParams( ));;}
}
export class TypeScriptImport {
	serialize() : Node {return new MapNode( 0).withNodeListAndSerializer( 0, 0.values, 0.serialize).withNodeListAndSerializer( 0, 0.segments, 0.serialize);;}
}
export class TypescriptRoot {
	serialize() : Node {return new MapNode( 0).withNodeListAndSerializer( 0, 0.children, 0.serialize);;}
}
export class TypescriptMethod {
	serialize() : Node {break;return 0.maybeChildren.map( 0).orElse( 0);;}
}
export class TypeScriptTemplateType {
	TypeScriptTemplateType(base : JavaLang.JavaSymbol, typeArguments : TypeArguments<TypeScriptType>) : public {break;break;;}
	serialize() : Node {return new MapNode( 0).withNodeSerialized( 0, 0.base).withNodeListSerialized( 0, 0.typeArguments.arguments( ));;}
}
export class TypescriptConditional {
	TypescriptConditional(type : ConditionalType, condition : TypescriptValue) : public {0( 0, 0);;}
	serialize() : Node {return new MapNode( 0.type.name( ).toLowerCase( )).withNodeSerialized( 0, 0.condition);;}
}
export class TypescriptBlock {
	TypescriptBlock(header : TypescriptBlockHeader, segments : List<TypescriptFunctionSegment>) : public {0( 0, 0);;}
	serialize() : Node {return new MapNode( 0).withNodeSerialized( 0, 0.header).withNodeListSerialized( 0, 0.children);;}
}
export class TypescriptConstructor {
	serialize() : Node {return new MapNode( 0);;}
}
export class TypeScriptDefinition {
	TypeScriptDefinition(definition : CommonLang.Definition<TypeScriptType>) : public {break;;}
	serialize() : Node {return new MapNode( 0).withString( 0, 0.definition.name( )).withNodeSerialized( 0, 0.definition.type( ));;}
}
export class TypescriptArrayType {
	TypescriptArrayType(arrayType : TypeScriptType) : public {break;;}
	serialize() : Node {return new MapNode( 0).withNodeSerialized( 0, 0.childType);;}
}
export class TypeParam {
	deserialize(node : Node) : CompileResult<TypeParam> {return 0.destruct( 0).withString( 0).complete( 0.new);;}
	serialize() : Node {return new MapNode( ).withString( 0, 0.value);;}
}
export class Construction {
	serialize() : Node {return new MapNode( 0).withNodeSerialized( 0, 0.type);;}
}
export class Post {
	serialize() : Node {return new MapNode( 0.variant.type( )).withNodeSerialized( 0, 0.value);;}
}
export class Access {
	serialize() : Node {return new MapNode( 0).withNodeSerialized( 0, 0.receiver).withString( 0, 0.property);;}
}
export class TypescriptBreak {
	serialize() : Node {return new MapNode( 0);;}
}
export class TypescriptContinue {
	serialize() : Node {return new MapNode( 0);;}
}
export class TypescriptReturnNode {
	TypescriptReturnNode(child : TypescriptValue) : public {0( 0);;}
	serialize() : Node {return new MapNode( 0).withNodeSerialized( 0, 0.child);;}
}
export class Invokable {
	Invokable(caller : TypescriptCaller, arguments : List<TypescriptArgument>) : public {0( 0, 0);;}
	serialize() : Node {return new MapNode( 0).withNodeSerialized( 0, 0.caller).withNodeListSerialized( 0, 0.arguments);;}
}
export class Char {
	serialize() : Node {return new MapNode( 0).withString( 0, 0.value);;}
}
export class Index {
	serialize() : Node {return new MapNode( 0).withNodeSerialized( 0, 0.parent).withNodeSerialized( 0, 0.argument);;}
}
export class Not {
	serialize() : Node {return new MapNode( 0).withNodeSerialized( 0, 0.child);;}
}
export class TypescriptLang {
}
